bundles/org.openhab.binding.ecovacs/src/main/java/org/openhab/binding/ecovacs/internal/handler/EcovacsVacuumHandler.java

```diff
@@ -0,0 +1,830 @@
+                }
+                break;
+            case CMD_STOP:
+                return new StopCleaningCommand();
+            case CMD_CHARGE:
+                return new GoChargingCommand();
+        }
+
+        if (command.startsWith(CMD_SPOT_AREA) && device.hasCapability(DeviceCapability.SPOT_AREA_CLEANING)) {
+            String[] splitted = command.split(":");
+            if (splitted.length == 2 || splitted.length == 3) {
+                int passes = splitted.length == 3 && "x2".equals(splitted[2]) ? 2 : lastDefaultCleaningPasses;
+                List<String> roomIds = new ArrayList<>();
+                for (String id : splitted[1].split(";")) {
+                    // We let the user pass in letters as in Ecovacs' app, but the API wants indices
+                    if (id.length() == 1 && id.charAt(0) >= 'A' && id.charAt(0) <= 'Z') {
+                        roomIds.add(String.valueOf(id.charAt(0) - 'A'));
+                    } else {
+                        logger.info("{}: Found invalid spot area room ID '{}', ignoring.", getDeviceSerial(), id);
+                    }
+                }
+                if (!roomIds.isEmpty()) {
+                    return new SpotAreaCleaningCommand(roomIds, passes);
+                }
+            } else {
+                logger.info("{}: spotArea command needs to have the form spotArea:<room1>[;<room2>][;<...roomX>][:x2]",
+                        getDeviceSerial());
+            }
+        }
+        if (command.startsWith(CMD_CUSTOM_AREA) && device.hasCapability(DeviceCapability.CUSTOM_AREA_CLEANING)) {
+            String[] splitted = command.split(":");
+            if (splitted.length == 2 || splitted.length == 3) {
+                String coords = splitted[1];
+                int passes = splitted.length == 3 && "x2".equals(splitted[2]) ? 2 : lastDefaultCleaningPasses;
+                String[] splittedAreaDef = coords.split(";");
+                if (splittedAreaDef.length == 4) {
+                    return new CustomAreaCleaningCommand(String.join(",", splittedAreaDef), passes);
+                }
+            }
+            logger.info("{}: customArea command needs to have the form customArea:<x1>;<y1>;<x2>;<y2>[:x2]",
+                    getDeviceSerial());
+        }
+
+        return null;
+    }
+
+    private interface WithDeviceAction {
+        void run(EcovacsDevice device) throws EcovacsApiException, InterruptedException;
+    }
+
+    private void doWithDevice(WithDeviceAction action) {
+        EcovacsDevice device = this.device;
+        if (device == null) {
+            return;
+        }
+        try {
+            action.run(device);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        } catch (EcovacsApiException e) {
+            logger.debug("{}: Failed communicating to device, reconnecting", getDeviceSerial(), e);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
```

Comment:
I would suggest to add a thing status detail here.

Reply:
Do you mean a description? (A thing status detail is already present)

Reply:
Added the description.